### ISSUE TITLE
PM: Link DMA fails to resume after system suspend

### ISSUE THREAD

**Issue by keqiaozhang** [thesofproject/linux#489]
This issue only happens with aplay, no such issue if use browser player to play audio via HDMI or I2S.
Plus, we only observe this issue on HDMI with 0.55 recipe.
But with PV recipe, both hdmi and i2s can reproduce this issue.

How to reproduce:
1. open one terminal and play audio via hdmi monitor.
    aplay -Dhw:0,5 -f dat -c 2 audio.wav
2. open another terminal and run suspend/resume test
    suspend_stress_test 
3. then system will reboot

No log is available.

PV recipe:
kernel: cros-sof-v4.14-rebase
sof/tplg: glk-005-drop-stable

**Comment by plbossart** [comment]
Good test!

IIRC correctly all streams are stopped by CRAS before suspend and explicitly restarted so the different behavior is "normal" since the flows are different. 

I believe you can force the same behavior with CRAS if you use the traditional method of suspending (echo mem > /sys/power/state), with CRAS using the RESUME_TRIGGER just like aplay. 

@ranj063 @cujomalainey can you double-check if my understanding is correct?

**Comment by mengdonglin** [comment]
@keqiaozhang Please help  check if this is the the same issue of https://github.com/thesofproject/sof/issues/788 reported on APL.

**Comment by keqiaozhang** [comment]
@mengdonglin 
I tried to disable system reboot when kernel panic on GLK, but system just died after running the test, still no useful log or trace available.

**Comment by keyonjie** [comment]
> @keqiaozhang Please help check if this is the the same issue of [thesofproject/sof#788](https://github.com/thesofproject/sof/issues/788) reported on APL.

I believe so, this is known issue.

**Comment by cujomalainey** [comment]
Following up with the CRAS team.

**Comment by ranj063** [comment]
@mengdonglin @keqiaozhang I've edited the title of this issue to be more indicative of what the problem is and I'm elevating its status to P1 because this needs to be fixed ASAP.

Basically we can resume host dma after system suspend but link DMA fails to resume.
@RanderWang @libinyang are both looking to fix it.

**Comment by cujomalainey** [comment]
@plbossart you are correct in your understanding

**Comment by libinyang** [comment]
@RanderWang should be looking into this bug? I'm working on 0x60000000 issue and suppose it is fixed by Ranjani's patch. @RanderWang any findings?

I'm currently assigned a task of letting sof-dev work on chrome.

**Comment by RanderWang** [comment]
I trace it today and don't find anything wrong in FW now. Now I go back to host driver and find that our host driver only receives stop and resume cmd in front end.  Back end dai never receive these cmds.

**Comment by ranj063** [comment]
@RanderWang please check out my explanation in PR #582. BE dai link do not have ops set for their PCM's.

**Comment by wenqingfu** [comment]
>@RanderWang @libinyang are both looking to fix it.

> @RanderWang please check out my explanation in PR #582. BE dai link do not have ops set for their PCM's.

So is this the same problem #582 try to fix?  @ranj063, will you continue to own this bug? or should @RanderWang own it?   Do we need more than one person on this?

**Comment by ranj063** [comment]
@wenqingfu no this isnt the same problem as #582. @RanderWang should own this as he is already working on a fix.

**Comment by libinyang** [comment]
> I trace it today and don't find anything wrong in FW now. Now I go back to host driver and find that our host driver only receives stop and resume cmd in front end. Back end dai never receive these cmds.

@RanderWang 
This is correct behavior as everything is done in FE. Although this is not designed very logically, it's efficient.

**Comment by RanderWang** [comment]
@ranj063  hi, I make great progress and update the detail in https://github.com/thesofproject/linux/pull/582.  For I will have a annual leave tomorrow, can you help to fix this bug? 

Thanks!

**Comment by mengdonglin** [comment]
@ranj063 ,

You will continue working on this issue ?I
You commented that "I will submit a separate PR for the link DMA resuming part" in https://github.com/thesofproject/linux/pull/582#issuecomment-459499043

I've checked with @RanderWang  and he has not been working on this.

**Comment by ranj063** [comment]
@mengdonglin yes but this is not a P1 issue. Could you please downgrade it to P2?

**Comment by lgirdwood** [comment]
@mengdonglin this is still showing in daily test. Do we have an ETA for fixing ?This is P1 for me sine aplay is gold standard for Linux userspace API usage.

**Comment by libinyang** [comment]
@lgirdwood 
https://github.com/thesofproject/linux/pull/725
is submitted. Ranjani suggestion we should implement the trigger order in FE link. As you know, if we implement in FE link, this will touch the ABI, alsa-lib, m4 topology file. It will take a longer time. I'm now implement it in FE link.
Another patch in hdmi codec is submitted in alsa mail list. I proposed 2 solutions, but Takashi's doesn't like it. Takashi proposed 4 solutions. I'm now submitting the patch for his 4th solution and wait for his review. And as per Pierre's suggestion, I will test the patch on cAVS tomorrow.

**Comment by kv2019i** [comment]
@libinyang Can you update on the status? This was discussed in the bugscrub today, but it's hard to understand the status from this list. PR824 is closed, what remains?

**Comment by libinyang** [comment]
@kv2019i Patch hw_params for BE is merged. However, we still need Ranjani's patch: https://github.com/thesofproject/linux/pull/880

So the bug still exists. After #880 is merged, this bug can be closed.

**Comment by libinyang** [comment]
Patch #880 is merged yesterday. I think we can close this bug now.
